tests/test_gmail_utils.py

```diff
@@ -0,0 +1,112 @@
+
+load_dotenv()
+
+def login_to_gmail(gmail, username, password):
+    "Login to mail"
+    result_flag = gmail.login(username, password)
+    if not result_flag:
+        raise RuntimeError("Login failed!")
+    print("Login successful!")
+    return 1, 1
+
+def fetch_and_print_mailboxes(gmail):
+    "fetch mailboxes the get the mailbox names"
+    mailboxes = gmail.fetch_mailboxes()
+    if not mailboxes:
+        raise ValueError("Failed to fetch mailboxes!")
+    print(f"Fetched mailboxes: {mailboxes}")
+    return 1, 1
+
+def fetch_and_print_messages(inbox_mailbox):
+    "fetch and get the messages"
+    messages = inbox_mailbox.mail()
+    print(f"Number of messages in SPAM: {len(messages)}")
+    return 1, 1, messages
+
+def fetch_subjects(gmail, messages):
+    "fetch the mails and get the subjects"
+    messages_dict = {msg.uid.decode('utf-8'): msg for msg in messages}
+    fetched_messages = gmail.fetch_multiple_messages(messages_dict)
+    if not fetched_messages:
+        raise RuntimeError("Failed to fetch multiple messages!")
+    print(f"Fetched multiple messages: {fetched_messages}")
+
+    for uid, message in fetched_messages.items():
+        subject = getattr(message, 'subject', 'No subject attribute')
+        print(f"UID: {uid}, Subject: {subject}")
+    return 1, 1
+
+@pytest.mark.GMAIL
+def test_gmail_util():
+    "Run the test"
+    expected_pass = 0
+    actual_pass = 0
+
+    try:
+        gmail = Gmail()
+        gmail.connect()
+
+        username = os.getenv('app_username')
+        password = os.getenv('app_password')
+
+        expected, actual = login_to_gmail(gmail, username, password)
+        expected_pass += expected
+        actual_pass += actual
+
+        expected, actual = fetch_and_print_mailboxes(gmail)
+        expected_pass += expected
+        actual_pass += actual
+
+        inbox_mailbox = gmail.use_mailbox('[Gmail]/Spam')
+        if not isinstance(inbox_mailbox, Mailbox):
+            raise TypeError(f"Error: Expected Mailbox instance, got {type(inbox_mailbox)}.")
+        print("SPAM selected successfully!")
+        expected_pass += 1
+        actual_pass += 1
+
+        expected, actual, messages = fetch_and_print_messages(inbox_mailbox)
+        expected_pass += expected
+        actual_pass += actual
+
+        if messages:
+            msg = messages[0]
+            fetched_msg = msg.fetch()
+            print(f"Fetching Message subject from test script: {fetched_msg.get('subject')}")
+            expected_pass += 1
+            actual_pass += 1
+
+            expected, actual = fetch_subjects(gmail, messages)
+            expected_pass += expected
+            actual_pass += actual
+        else:
+            print("No messages found in SPAM.")
+
+    except (TypeError, ValueError, KeyError) as e:
+        print(f"Exception when trying to run test: {__file__}")
+        print(f"Python says {str(e)}")
+
+    finally:
+        gmail.logout()
+        print("Logged out!")

```

Comment:
If the login itself fails, the logout is called?
Also we dont see any failure for logout

Reply:
We need to print some message for login failure
tests/test_gmail_utils.py::test_gmail_util Logged out!
FAILED

================================================================================================= FAILURES =================================================================================================
_____________________________________________________________________________________________ test_gmail_util ______________________________________________________________________________________________
tests/test_gmail_utils.py:67: in test_gmail_util. Currently we see this error
    expected, actual = login_to_gmail(gmail, username, password)
tests/test_gmail_utils.py:21: in login_to_gmail
    result_flag = gmail.login(username, password)
integrations/reporting_channels/gmail/gmail.py:123: in login
    imap_login = self.imap.login(self.username, self.password)
/usr/lib/python3.10/imaplib.py:610: in login
    typ, dat = self._simple_command('LOGIN', user, self._quote(password))
/usr/lib/python3.10/imaplib.py:1222: in _quote
    arg = arg.replace('\\', '\\\\')
E   AttributeError: 'NoneType' object has no attribute 'replace'

Reply:
- moved the logout() inside the try block and also updated the login() to print the error message when the login fails.

`$ python -m pytest test_gmail_utils.py
============================= test session starts =============================
platform win32 -- Python 3.11.2, pytest-8.1.1, pluggy-1.4.0 -- C:\Users\mohan\code\qxf2\venv\Scripts\python.exe
metadata: {'Python': '3.11.2', 'Platform': 'Windows-10-10.0.22631-SP0', 'Packages': {'pytest': '8.1.1', 'pluggy': '1.4.0'}, 'Plugins': {'anyio': '3.6.2', 'html': '4.1.1', 'metadata': '3.1.1', 'reportportal': '5.4.0', 'rerunfailures': '14.0', 'snapshot': '0.9.0', 'xdist': '3.5.0'}, 'JAVA_HOME': 'C:\\Program Files\\Java\\jdk-22'}
rootdir: C:\Users\mohan\code\qxf2\qxf2-page-object-model
configfile: pytest.ini
plugins: anyio-3.6.2, html-4.1.1, metadata-3.1.1, reportportal-5.4.0, rerunfailures-14.0, snapshot-0.9.0, xdist-3.5.0
collecting ... collected 1 item

test_gmail_utils.py::test_gmail_util Login failed due to invalid credentials.
FAILEDError processing consolidated log file: [Errno 2] No such file or directory: 'log\\consolidated_log.txt'


================================== FAILURES ===================================
_______________________________ test_gmail_util _______________________________
test_gmail_utils.py:78: in test_gmail_util
    expected, actual = fetch_and_print_mailboxes(gmail)
test_gmail_utils.py:35: in fetch_and_print_mailboxes
    mailboxes = gmail.fetch_mailboxes()
..\integrations\reporting_channels\gmail\gmail.py:63: in fetch_mailboxes
    response, mailbox_list = self.imap.list()
..\..\..\..\AppData\Local\Programs\Python\Python311\Lib\imaplib.py:599: in list
    typ, dat = self._simple_command(name, directory, pattern)
..\..\..\..\AppData\Local\Programs\Python\Python311\Lib\imaplib.py:1230: in _simple_command
    return self._command_complete(name, self._command(name, *args))
..\..\..\..\AppData\Local\Programs\Python\Python311\Lib\imaplib.py:968: in _command
    raise self.error("command %s illegal in state %s, "
E   imaplib.IMAP4.error: command LIST illegal in state NONAUTH, only allowed in states AUTH, SELECTED
`